File: Predictive_parser.c
#include <stdio.h>
#include <string.h>

char stack[50];
int top = -1;

void push(char c) {
    stack[++top] = c;
}

char pop() {
    return stack[top--];
}

int main() {
    char input[50];
    int i = 0;

    printf("Enter input string (end with $): ");
    scanf("%s", input);

    push('$');
    push('S'); // Start symbol

    while(top != -1) {
        char stackTop = stack[top];
        char current = input[i];

        // If match
        if(stackTop == current) {
            pop();
            i++;
        }
        // Example grammar: S → aSb | ε
        else if(stackTop == 'S' && current == 'a') {
            pop();
            push('b');
            push('S');
            push('a');
        }
        else if(stackTop == 'S' && current == 'b') {
            pop(); // ε production
        }
        else if(stackTop == '$' && current == '$') {
            printf("String Accepted\n");
            return 0;
        }
        else {
            printf("String Rejected\n");
            return 0;
        }
    }

    return 0;
}

File: Predictive_parser.java
import java.util.*;

public class PredictiveParser {

    static Stack<Character> stack = new Stack<>();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter input string (end with $): ");
        String input = sc.next();
        int i = 0;

        stack.push('$');
        stack.push('S');

        while(!stack.isEmpty()) {
            char top = stack.peek();
            char current = input.charAt(i);

            // Match terminal
            if(top == current) {
                stack.pop();
                i++;
            }
            // Grammar: S → aSb
            else if(top == 'S' && current == 'a') {
                stack.pop();
                stack.push('b');
                stack.push('S');
                stack.push('a');
            }
            // S → ε
            else if(top == 'S' && current == 'b') {
                stack.pop();
            }
            else if(top == '$' && current == '$') {
                System.out.println("String Accepted");
                return;
            }
            else {
                System.out.println("String Rejected");
                return;
            }
        }

        sc.close();
    }
}